feat: add odin mode

laytan opened the pull request:
*Issue #, if available:* N/A

*Description of changes:* Added a basic mode for Odin


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

akoreman commented:
Thanks for your contribution, could you also add an example doc for the mode to the `docs/kitchen-sink/docs` folder and add it to the modelist in `src/ext/modelist.js`. For reference on what to add where, see PR #5037.

laytan commented:
@akoreman I have added the things you asked :)

akoreman commented:
Thanks :) I've been comparing your mode with the syntax highlighting found in the [Odin docs](https://odin-lang.org/docs/overview/) examples and it seems to match that pretty well.

One small quirk I noticed was highlighting for names which start with a capital (see picture, e.g. Vector3):

<img width="264" alt="Screenshot 2023-05-14 192913" src="https://github.com/ajaxorg/ace/assets/34573235/ec8ca6c0-afca-40f2-b4d8-76ca96c255ca">

I don't know enough about Odin to tell what the desired behaviour is here but could you take a look at this?

Could you also add a test covering the indent/outdent behavior (for an example of such tests, see e.g. [here](https://github.com/ajaxorg/ace/blob/master/src/mode/ada_test.js)).

laytan commented:
@akoreman I've added the tests, let me know if they are ok like this.

Also removed the uppercase === constant rule because of the issue you were having and also that it can cause false-positives anyway.

akoreman commented:
Thanks for adding the tests, they look good to me. 

Correct me if I'm wrong, but I think you might have removed the wrong lines from the highlight file. AFAIK the lines you removed are for `#xyz_abc` (which are [Odin directives](https://odin-lang.org/docs/overview/#directives) if I read correctly). I think you meant to remove [these lines](https://github.com/ajaxorg/ace/pull/5169/files#diff-b400de2ec23b069012324dfbb7b5c558710e0c525f6fbc4ab56ded204cea2782R135-R138)?

valerio-123 reviewed (commented):
Mamma

valerio-123 reviewed (commented):
Mamma
